### DISCUSSION TITLE
automatic refetching only on key change (no refetch of stale data)

### DISCUSSION THREAD

**Discussion by dominikdosoudil** [TanStack/query#8306] [Q&A] [answered]
Use case:
- User wants to refetch data manually (click) so the list items do not change too often and he is able to read it.

The idea:
- Disable query so it does not change itself,
- invalidate query when I know that there are new records (websocket event),
- let the user know that he can refetch showing him a button `{query.isStale ? <button ... /> : null}`

Unfortunately I expected that isStale would still change even if the query is disabled but it does not. Refetching the disabled query works tho. Is there a way how to make it work? 

`@tanstack/react-query@5.60.6`

Code snippets with my configuration:

```typescript
  queryOptions({
    queryKey: ['foo', page],
    queryFn: async () => {
      ...
    },
    retry: false,
    placeholderData: keepPreviousData,
    staleTime: Infinity,
    enabled: false,
  });

```

```typescript
     queryClient.invalidateQueries({
        queryKey: ['foo'],
        type: 'all',
      });
```

EDIT:

Actually, changing page won't work as well. I've got confused while trying different options.
Anyway, I guess that I need something like: `refetchOnStale: false` but I couldn't found anything like that in the docs.

**Comment by TkDodo** [comment]
Funny because I’ve just implicitly addressed it in this PR:

- #8313 

I used to make a fix that said that disabled queries shouldn’t be shown as `stale`, but I think that was wrong. If we ship the above PR, this should be fixed.

**Comment by dominikdosoudil** [comment]
@TkDodo Wow, that's cool. I believe this would be really helpful in my scenario. Thanks

However I will still be unable to acquire the behaviour "Refetch on key change, do not refetch when it becomes stale" am I right?

**Reply by TkDodo** [reply, marked as answer]
staleness defines when a smart refetch should happen, so no. But you can turn off all the refetchOn... triggers and use StaleTime to just let the user know that data isn't fresh with the button

**Reply by dominikdosoudil** [reply]
There is the thing that there is also few filters. I would need to somehow make sure that the refetch fn is also called after the filter is selected. I think that I will give it a try but right now it seems to me that RQ is just not the right tool for this exact use case. However thank you very much anyway, it will be useful in other use cases of mine for sure.
